usb/dev.c

```diff
@@ -232,19 +231,44 @@ static int usb_getConfiguration(usb_dev_t *dev)
 	}
 
 	ptr = (char *)conf + sizeof(usb_configuration_desc_t);
-	for (i = 0; i < dev->nifs; i++) {
-		iface = (usb_interface_desc_t *)ptr;
-		ptr += sizeof(usb_interface_desc_t);
-		/* Class and Vendor specific descriptors */
-		/* TODO: save them to allow drivers access */
-		while (ptr[1] != USB_DESC_ENDPOINT && ptr < (char *)conf + pre.wTotalLength)
-			ptr += ptr[0];
-		if (ptr >= (char *)conf + pre.wTotalLength)
+	size = pre.wTotalLength - sizeof(usb_configuration_desc_t);
+	nifs = 0;
+
+	while ((size >= sizeof(struct usb_desc_header)) && ((size -= ((struct usb_desc_header *)ptr)->bLength) >= 0)) {
+		if (((struct usb_desc_header *)ptr)->bLength < sizeof(struct usb_desc_header))
 			break;
-		dev->ifs[i].eps = (usb_endpoint_desc_t *)ptr;
-		dev->ifs[i].desc = iface;
-		while (ptr < (char *)conf + pre.wTotalLength && ptr[1] == USB_DESC_ENDPOINT)
-			ptr += sizeof(usb_endpoint_desc_t);
+
+		switch (((struct usb_desc_header *)ptr)->bDescriptorType) {
+			case USB_DESC_INTERFACE:
+				if (((struct usb_desc_header *)ptr)->bLength < sizeof(usb_interface_desc_t))
+					break;
+				if (++nifs <= dev->nifs)
+					dev->ifs[nifs - 1].desc = (usb_interface_desc_t *)ptr;
+				break;
+
+			case USB_DESC_ENDPOINT:
+				if (((struct usb_desc_header *)ptr)->bLength < sizeof(usb_endpoint_desc_t))
+					break;
+				if (nifs && (nifs <= dev->nifs) && (dev->ifs[nifs - 1].eps == NULL))
```

Comment:
violates MISRA Rule 14.4 (The controlling expression of an if statement and the controlling expression of an iteration-statement shall have essentially Boolean type)

we need to write `nifs != 0`  explicitly :/